### PR TITLE
V0.3.0 - Metric Function & Overhaul

### PR DESCRIPTION
**Pull request by callum-mcdata**
## What Does This PR Do?
Honestly, it's probably a better question to ask what **doesn't** this PR do. This PR is a complete overhaul and restructure of the metrics package to better integrate it with the new `metric` function that @drewbanin added to @dbt-labs/core . But for posterity's sake, lets break this out as best we can.

### Components
1. Changing the `metric_name` input to instead support `metric('metric_name')`
2. Query restructuring
   - The current query structure was great but unfortunately it didn't scale well with multiple metrics. We did significant testing internally and determined that this newly implemented structure was the most performant with larger datasets.
3. Adding support for `expression` metrics. We'll probably rename these later on but the functionality is there.
4. Adding testing for multiple components (start date, end date, secondary calculations, expression metrics)
5. Breaking everything down into components

### CODE REVIEW

**Comment by callum-mcdata** [comment]
https://www.loom.com/share/4aeeb94c379643e39e4ac2b99cc5ab25

**Comment by callum-mcdata** [comment]
https://www.loom.com/share/4aeeb94c379643e39e4ac2b99cc5ab25

File: integration_tests/models/custom_calendar.sql [hunk 1]
```diff
@@ -20,7 +20,11 @@ final as (
         {% endif %}
         cast({{ dbt_utils.date_trunc('month', 'date_day') }} as date) as date_month,
         cast({{ dbt_utils.date_trunc('quarter', 'date_day') }} as date) as date_quarter,
-        cast({{ dbt_utils.date_trunc('year', 'date_day') }} as date) as date_year
+        cast({{ dbt_utils.date_trunc('year', 'date_day') }} as date) as date_year,
+        case 
+            when dayname(to_date(date_day)) in ('Sat','Sun') then true
+            else false 
+        end as is_weekend
```
Comment (dbeatty10) [inline integration_tests/models/custom_calendar.sql:27]: Surprised that `is_weekend` isn't defined in some cross-database macro somewhere!

We should probably get that defined so this functionality works cross-database.

Comment (callum-mcdata) [reply integration_tests/models/custom_calendar.sql:27]: It very well could be! To be honest I only added this field to test out functionality around bringing in dimensions from calendar tables.

**Comment by callum-mcdata** [comment]
LET THERE BE PRERELEASE!

**Pull request merged**